### ISSUE TITLE
[BUG] Native C++ Callback does not work when running Electron App with Node Add-on using Playwright

### ISSUE THREAD

**Issue by anvosoughi** [microsoft/playwright#12569]
**Context:**
- Playwright Version: 1.19.2
- Operating System: Windows 10
- Node.js version: 16.14.0
- Electron: 17.1.0 
- Node Add-on: build with Electron "17.1.0" and "node-addon-api": "^4.3.0"

**Bug Repository**

https://github.com/anvosoughi/playwright-electron-node-addon-sample

**Electron Code:**

https://github.com/anvosoughi/playwright-electron-node-addon-sample/blob/main/electron/main.js

**Playwright Code**

https://github.com/anvosoughi/playwright-electron-node-addon-sample/blob/main/e2e/specs/home.spec.ts

**Describe the bug**

I have a simple Electron sample app, which uses the 'node-addon-api' to make a call from JavaScript(electron app) to C++ add-on, and receive a callback from C++ add-on to the JavaScript(electron app).

When the app is launched, the triggerCallback function is automatically executed, which triggers C++ add-on to make one call to the app every second using a callback. 

There is a console output of "triggerCallback fired" when the function is executed. And "callback result:  XX" is printed to console every second thereafter.

**This works well when launching the app without Playwright** 
![image](https://user-images.githubusercontent.com/81978600/157131024-70404c04-d0c8-48bb-b253-395fc3257277.png)

**But when we use Playwright to launch the app, the function is executed as expected, but the call back does not fire.**
![image](https://user-images.githubusercontent.com/81978600/157130982-92684295-ffb3-4822-9d71-b19d17f4c24f.png)

**Note:** 

You can also trigger the callback manually using the button with text "Trigger Callback"

This app is for Windows OS only

**Comment by yury-s** [comment]
I cannot reproduce it on Linux, with modified preload.js:

```js
const loadLib = () => {
  if (PLATFORM === "win32") {
    return require(`./libs/Test_Module_Windows.node`);
  }

  return require(`./libs/Test_Module_Linux.node`);
};
```
and it worked well:

![image](https://user-images.githubusercontent.com/9798949/157135298-f7c445da-6ef1-4a60-a455-532bd9f2a01f.png)

Perhaps the library path doesn't work well on Windows?

**Comment by anvosoughi** [comment]
@yury-s thanks for taking a look. But the issue we have is only when launching the app with Playwright, and only on Windows.

Can you confirm if you tested with running app with Playwright? ie. "npm run test:e2e" ?

**Comment by yury-s** [comment]
> Can you confirm if you tested with running app with Playwright? ie. "npm run test:e2e" ?

Yes, this is how I launched it. I tested it only on linux though.

> Also not sure what you mean by wrong path? because the app works without Playwright...

If it works without playwright it rules out my hypothesis. Can you add debug logging to see if  "return require(`./libs/Test_Module_Windows.node`);" runs without errors? I'd also suggest you run with `DEBUG=pw:browser` environment variable to see the browser output, it may give you a hint.

**Comment by anvosoughi** [comment]
@yury-s Here is the output with pw:browser and pw:api. Nothing stands out to me. Although everything passes, the call backs are still not working when looking at the console output, and this is again only when launching app through Playwright.

pw:api
![image](https://user-images.githubusercontent.com/81978600/157315386-10739421-2c09-422a-957f-97c503e3e959.png)

pw:browser
![image](https://user-images.githubusercontent.com/81978600/157315445-fff006a9-7ecf-4e41-856e-14092d292722.png)

**Comment by yury-s** [comment]
The output looks normal. One difference is that I ran the test with npm but I don't think it matters. It looks like `helloWorld.testModule.trigger_callback();` is executed well (there is no exception in the console), did you try inserting logs into the native module to see where the execution stalls? It's hard to tell why the code is not triggered properly under playwright, playwright just starts electron as a regular binary, compared to manual run we just pass the ` --inspect=0 --remote-debugging-port=0` command line arguments.

**Comment by anvosoughi** [comment]
@yury-s 

I updated my "start" command to be "electron ./electron/main.js --inspect=0 --remote-debugging-port=0"

I added lots of logs in the native module. I found two differences between using Playwright and not.

1. 
What I found was that require() is only called once as expected on JS side, but on native, the node thread calls load_API() 2 times when running with Playwright. But only once without Playwright, when we use "start" command.

2.
When using Playwright, after we call NonBlockingCall() in our native thread, the JS callback "javascript_callback()" is not fired....
But is called a single time at the very end of the log, when the app is being shutdown...
When not using Playwright, after we call NonBlockingCall() in our native thread, the JS callback "javascript_callback()" is fired every second as expected

The log files for attached for both scenarios:
[native-electron.log](https://github.com/microsoft/playwright/files/8218210/native-electron.log)
[native-playwright.log](https://github.com/microsoft/playwright/files/8218211/native-playwright.log)

**Comment by pavelfeldman** [comment]
Electron support is still experimental in Playwright, so we are not prioritizing it. If you nail down the bug and have a fix, we'll be happy to review and accept it.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

We are prioritizing the bugs based on the upvotes, recency and our ability to act. It looks like this issue only has a handful of upvotes, has not been touched recently and/or we lack sufficient feedback to act on it. We are closing issues like this one to keep our bug database maintainable. Please feel free to open a new issue and link this one to it if you think this is a mistake.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your contribution to our project. This issue has been closed due to its limited upvotes and recent activity, and insufficient feedback for us to effectively act upon. Our priority is to focus on bugs that reflect higher user engagement and have actionable feedback, to ensure our bug database stays manageable.

Should you feel this closure was in error, please create a new issue and reference this one. We're open to revisiting it given increased support or additional clarity. Your understanding and cooperation are greatly appreciated.
